src/cms/forms/media/media_file_form.py

```diff
@@ -23,4 +24,23 @@ class Meta:
         fields = (
             "name",
             "alt_text",
+            "file",
+            "thumbnail"
         )
+
+    def clean(self):
+        """
+        Validate form fields which depend on each other, see :meth:`django.forms.Form.clean`:
+        Don't allow multiple root nodes for one region:
+        If the file type is invalid, add a :class:`~django.core.exceptions.ValidationError`.
+
+        :return: The cleaned form data
+        :rtype: dict
+        """
+        
+        cleaned_data = super().clean()
+        file = cleaned_data.get("file")
+        if file:
+            cleaned_data["thumbnail"] = generate_thumbnail(file, 300, 300, False)
+
+        return cleaned_data
```

Comment:
This is the right idea - however, we need a bit more complex validation and save methods for replacement, therefore I suggest to use a whole new form class for this task. So the `MediaFileForm` should be unchanged, but you can create a new `ReplaceMediaFileForm` with copy + paste from `UploadMediaFileForm`.

At first, we need to specify the fields for this form - since the `parent_directory` and `type` should be identical, we can remove it from the `fields` list.

Then, in the `clean()`-method, we have to do the following:
- Does the actual file type of the new file (checked with `magic.from_buffer()`) match the original file type? If not, throw a `ValidationError`. We only want to allow the replacement with new files of the same type.
- Set `cleaned_data["name"] = file.name` to make sure the new name in the database is the name of the new file
- Set `file.name = self.instance.file.name` to make sure the new name on the hard drive is the name of the old file
- If everything goes well and no errors occur, generate a thumbnail with the new file
- Set the thumbnail filename to the existing thumbnail filename of the form instance

Then, in the save-method, we have to do the following:
- Delete the old file with `self.instance.file.delete()` (otherwise Django would automatically append a random string to the new file's name because the name is not available)
- Delete the old thumbail
- Call `super().save()` to save the new files

One pitfall we have to keep in mind: the [upload_path()](https://github.com/Integreat/integreat-cms/blob/3632dd978cac673809cb72077aee1dc80c4c774b/src/cms/models/media/media_file.py#L24-L45)-method will return the directory of the current date instead of the initial file's date, so we have to find a way around this. However, I think it will be easier to tackle this problem when everything else is working (so we can try out a few different things), so let's not worry about that now :wink:

Reply:
Thanks for your advice! Most of your ideas seem to work flawless. However, the `self.instance.file.delete()` with in the save-method seems not to work and I couldn't really figure out why. Instead I resolved it by "manually" deleting the file with `os.remove()` and getting the correct path from the `models.media.media_file.upload_path()`-method.